### PR TITLE
API: Remove integer position args from xy for plotting

### LINKED ISSUE #20056: BUG: df.plot fails when given x,y args as positions
#### Code Sample, a copy-pastable example if possible

```python
from pandas import DataFrame

df = DataFrame({"A": [1, 2], 'B': [3, 4]})
df.plot(x=0, y=1)
```
#### Problem description

```
IndexError: index 1 is out of bounds for axis 0 with size 1
```

I came across this when writing test cases for my PR #20000 - looks like it happens due to calling `set_index` on the data rather than on a copy. I can potentially fix it there or open a separate PR.

### PR DESCRIPTION
**Pull request by masongallo**
- [x] closes #20056 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [?] whatsnew entry

This is based on discussion from #20000 where we decided it was confusing to allow `df.plot` to support both labels AND positions. I wasn't sure where to put this in whatsnew, so lmk and I'll update it.

### CODE REVIEW

File: pandas/tests/plotting/test_frame.py [hunk 1]
```diff
@@ -1054,17 +1050,15 @@ def test_plot_scatter_with_c(self):
...
+        axes = df.plot.scatter(x='x', y='y', c='z')
 
-            if self.mpl_ge_1_3_1:
+        # default to Greys
+        assert axes.collections[0].cmap.name == 'Greys'
 
-                # n.b. there appears to be no public method to get the colorbar
-                # label
-                assert ax.collections[0].colorbar._label == 'z'
+        if self.mpl_ge_1_3_1:
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: @TomAugspurger we should be able to blow this code away as we > 1.4.3, yes (separate PR)

File: pandas/tests/plotting/test_frame.py [hunk 2]
```diff
@@ -2172,14 +2166,27 @@ def test_invalid_kind(self):
...
+        (0, 1)
     ])
     def test_invalid_xy_args(self, x, y):
-        # GH 18671
+        # GH 18671 and # GH 20056
         df = DataFrame({"A": [1, 2], 'B': [3, 4], 'C': [5, 6]})
         with pytest.raises(ValueError):
             df.plot(x=x, y=y)
 
+    @pytest.mark.parametrize("x,y,colnames", [
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: can you add a test that checks the raise when passing ints and a non-int index

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a sub-section in the whatsnew about this in api-breaking changes. Note we *may* want to consider deprecating this and removing in a future version @jorisvandenbossche @TomAugspurger . though I am fine with just changing it.

**Comment by jorisvandenbossche** [comment]
> Note we may want to consider deprecating this and removing in a future version @jorisvandenbossche @TomAugspurger . though I am fine with just changing it.

If it worked, I think we should deprecate it. But did it actually work? Because https://github.com/pandas-dev/pandas/issues/20056 seems to indicate not? (if it didn't work, of course fine to just remove)

**Comment by jorisvandenbossche** [comment]
So it seems that for `line`, `bar`, `area` it doesn't work, but for `scatter`, `pie` and `hexbin` it does?

**Comment by TomAugspurger** [comment]
I don't think we should remove this. I know positional plotting gets some use, especially for exploratory analysis when you have long column names.

I'll look more at why the original issue was failiing.

**Comment by MasonGallo** [comment]
> So it seems that for line, bar, area it doesn't work, but for scatter, pie and hexbin it does?

@jorisvandenbossche yes I believe this is due to the structure of the code - the index setting I mentioned in #20056 only happens to `kind` not from that set - called `_dataframe_kinds` and `_series_kinds`.

> I don't think we should remove this. I know positional plotting gets some use, especially for exploratory analysis when you have long column names.

@TomAugspurger why not ask the user to supply `data.columns[ind]` instead of us making assumptions and calculating it for them? IMO that would make the API less complex.

**Comment by TomAugspurger** [comment]
IME, DataFrame.plot is most often used in exploratory analysis. For this,
speed typing becomes an API consideration.
If you have a long column name, but want to quickly plot it, `df.plot(y=0)`
is nicer than `df.plot(y='percent_return_for_foo_over_bar')`.

> instead of us making assumptions

I don't think we ever make assumptions? If there's ever an integer in the
column we don't fall back to positional indexing.
IIUC there was just a bug in how we computed the positions when both `x`
and `y` are specified.

**Comment by MasonGallo** [comment]
> IME, DataFrame.plot is most often used in exploratory analysis. For this,
speed typing becomes an API consideration.

That's fair - you have much more context than I do on usage.

> I don't think we ever make assumptions?

We make an assumption by allowing mixed types in column names - i.e. we check with `holds_integer` and then assume a given int is meant as a column name.

Seems like we don't have consensus on whether this should be removed from API. How should we proceed? I can close this and fix the integer indexing bug in #20000 ?

**Comment by TomAugspurger** [comment]
> We make an assumption by allowing mixed types in column names - i.e. we check with holds_integer and then assume a given int is meant as a column name.

Perhaps it's just semantics then, but I wouldn't call that assuming :) To me the behavior is clear (but could be better documented). Specifying `x` and `y` as positions is allowed if and only if the columns holds no integers.

> I can close this and fix the integer indexing bug in #20000 ?

If you're comfortable with maybe wasting some effort, I would work around the bug in #20000. We can wait to hear what the other maintainers say before closing or working further on this issue, but I would be sad to see this go.

**Comment by jorisvandenbossche** [comment]
> Perhaps it's just semantics then, but I wouldn't call that assuming :) To me the behavior is clear (but could be better documented). Specifying x and y as positions is allowed if and only if the columns holds no integers.

Although there might be some "guessing" cases (mixed integer columns), this is also behaviour that used a lot throughout the code base (eg `level` can often be specified both as integer or as name).

--

I don't have a strong feeling about this. If it is not too difficult to fix this to actually work, I am certainly fine with keeping it.

**Comment by MasonGallo** [comment]
FYI: will be closing this in favor of #20000

**Pull request closed without merging**